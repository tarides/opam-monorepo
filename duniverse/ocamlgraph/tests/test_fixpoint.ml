
(* Test file for Fixpoint *)

let id x = x
module IntOrdered = struct
  type t = int
  let compare: int -> int -> int = compare
  let hash: int -> int = id
  let equal: int -> int -> bool = (=)
end
module IntSet = Set.Make(IntOrdered)
module G = Graph.Persistent.Digraph.Concrete(IntOrdered)
module Divisors = Graph.Classic.P(G)
module Analysis = struct
  type data = IntSet.t
  type edge = G.edge
  type vertex = G.vertex
  type g = G.t
  let direction = Graph.Fixpoint.Backward
  let join = IntSet.union
  let equal = IntSet.equal
  let analyze (_: edge): data -> data = id
end
module Fixpoint = Graph.Fixpoint.Make(G)(Analysis)

let pp_int_set pp set =
  let first = ref true in
  Format.fprintf pp "@[<hov>";
  IntSet.iter (fun x ->
      if !first then
	first := false
      else 
	Format.fprintf pp ",@ ";
      Format.pp_print_int pp x)
    set;
  Format.fprintf pp "@]"

let () =
  let n = 15 in
  let labels = Fixpoint.analyze IntSet.singleton (Divisors.divisors n) in
  Format.open_vbox 0;
  for i = 2 to n do
    Format.printf "Labels for %d: %a@," i pp_int_set (labels i)
  done;
  Format.close_box ();
  Format.print_flush ()

