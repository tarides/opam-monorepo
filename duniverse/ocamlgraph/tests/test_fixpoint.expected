Labels for 2: 2, 4, 6, 8, 10, 12, 14
Labels for 3: 3, 6, 9, 12, 15
Labels for 4: 4, 8, 12
Labels for 5: 5, 10, 15
Labels for 6: 6, 12
Labels for 7: 7, 14
Labels for 8: 8
Labels for 9: 9
Labels for 10: 10
Labels for 11: 11
Labels for 12: 12
Labels for 13: 13
Labels for 14: 14
Labels for 15: 15
